### PR TITLE
MNTOR-983, MNTOR-1057 settings v2 frontend and migrate verification email

### PR DESCRIPTION
**Pull request by rhelmer**
This covers the following issues:

- migrate email to new platform MNTOR-1057
- frontend work for v2 settings MNTOR-983

It depends on the backend changes in https://github.com/mozilla/blurts-server/pull/2743

### CODE REVIEW

**Comment by rhelmer** [comment]
@flozia a few things are missing in V2 that we need to implement this, that aren't explicitly covered in figma:

- error messages
  - we can either redirect to an error page (which is what V1 does), or we can show some kind of error message using JS: 
image
  - see [src/client/js/settings.js](https://github.com/mozilla/blurts-server/pull/2763/files#diff-790e0d9a5e177ba94933197ab52bc546ed75f9a5026f8e306572cc0b02dc3a0c) for where the client is `fetch`ing the API (and not handling the exceptions right now just logging them to the console)
- success messages
  - I didn't find this in figma, but I see that V1 (current production on monitor.firefox.com) shows this using JS: 
image
- image formats
  - I exported the images in use here from Figma using PNG, is SVG preferred?

File: locales/en/settings.ftl [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+comm-opt-0 = Email me if one of my email addresses below appears in a data breach.
+comm-opt-1 = Send all breach alerts to { $primaryEmail }
+
+send-verification = Send Verification Link
+add-new-email = Add a new email address
+email-verification-required = Email verification required
+settings-title = { -product-short-name } Settings
+monitored-email-addresses = Monitored email addresses
+email-address-limit = Your account includes monitoring of up to { $limit } emails
```
Comment (flodolo) [inline locales/en/settings.ftl:23]: This should use the plural structure (like the first string in this file)

**Review by flodolo — CHANGES_REQUESTED** [review]
See the linter results https://github.com/mozilla/blurts-server/pull/2763/checks

More important: what's the reasoning behind splitting these strings into a separate file? This will require retranslating them across all locales.

If we do this, we should relook at the IDs that we used, and use better ones (providing more context on how/where the string is used).

**Comment by toufali** [comment]
> what's the reasoning behind splitting these strings into a separate file?

Hi @flodolo – as part of this Refresh release, we're re-architecting/refactoring.  The strings @rhelmer moved over are specific only to the settings page, and we have been creating a new .ftl for each page in the site (~5 total expected for Refresh).

The app.ftl file has become very large and hard to navigate – We're hoping to reserve this page for site-wide strings only. I was under the impression that .ftl files concat and that moving strings between files was ok, but it sounds like I was wrong!  

For this case is it ok to separate into a more logical file structure?  Monitor is expected to continue growing, and I'd like to set up good practice in the Refresh release, so future features can benefit.

**Comment by flodolo** [comment]
> The app.ftl file has become very large and hard to navigate – We're hoping to reserve this page for site-wide strings only. I was under the impression that .ftl files concat and that moving strings between files was ok, but it sounds like I was wrong!

No concatenation ;-) 
https://pontoon.mozilla.org/it/firefox-monitor-website/

I'm OK with splitting things up in smaller files, as long as it's done purposefully. Pontoon will show a match in translation memory, so it doesn't need to be translated from scratch. But we should also take the opportunity to do a better job with IDs, since we're basically creating new strings.

I'm heading offline for the week, but I'll try to provide more comments in the next days of what I mean.

File: locales/en/settings.ftl [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# This string is shown beneath each of the user’s email addresses to indicate
+# how many known breaches that email address was found in.
+appears-in-x-breaches =
+  { $breachCount ->
+        [one] Appears in { $breachCount } known breach.
+       *[other] Appears in { $breachCount } known breaches.
+  }
+
+resend-verification = Resend verification email
```
Comment (flodolo) [inline locales/en/settings.ftl:9]: I understand these come from existing strings, but we didn't do a good job here, and we can use this work to improve things.

IDs should provide context for localizers, if there's no comment doing it. For example: is this a button, a link, something else? Putting more information, and using a more specific ID based on the context would help localizers, and discourage reuse in different contexts.

In this case I would go with `settings-resend-verification-link`.

This applies to all the strings in this file, in particular to very undescriptive ones like `menu`.
`email-verification-required`: is it a title, a message? These might require different style in different locales.

File: src/client/js/settings.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+      method: 'POST',
+      body: JSON.stringify({ email })
+    })
+
+    if (!response || !response.ok) {
+      throw new Error(`sending verification email failed: ${response.error}`)
+    }
+
+    const addEmailDialogContent = document.getElementById('add-email-modal-content')
+    addEmailDialogContent.textContent = `Verify the link sent to ${email} to add it to Firefox Monitor. Manage all email addresses in Settings.`
```
Comment (pdehaan) [inline src/client/js/settings.js:86]: Q: Does this need to be localized?

File: src/utils/email.test.js [hunk 4]
```diff
@@ -0,0 +1,234 @@
+'use strict'
+
+import {
+  TEST_SUBSCRIBERS,
+  TEST_EMAIL_ADDRESSES
+} from '../db/seeds/test_subscribers'
+import { initFluentBundles } from './fluent.js'
+
+import AppConstants from '../app-constants'
+import EmailUtils from './email'
```
Comment (pdehaan) [inline src/utils/email.test.js:10]: Should these have .js extensions for consistency?

File: src/views/email-2022.js [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  </td>
+</tr>
+`
+function getEmailFooterCopy (data) {
+  let faqLink = getMessage(
+    'frequently-asked-questions'
+  )
+  faqLink = `<a href="https://support.mozilla.org/kb/firefox-monitor-faq">${faqLink}</a>`
+
+  if (!(data.whichPartial === 'email_partials/email-monthly-unresolved')) {
```
Comment (pdehaan) [inline src/views/email-2022.js:67]: Why not this?
```suggestion
  if (data.whichPartial !== 'email_partials/email-monthly-unresolved') {
```

File: src/views/partials/settings.js [hunk 6]
```diff
@@ -1,5 +1,93 @@
...
+}
+
+export const settings = (data) => `
+<div id="settings" class="parent" data-csrf-token="${data.csrfToken}">
+  <div class="child-left">
+    <h2>${getMessage('settings-title')} </h2>
+  </div>
+  <div class="child-right">
+    <section>
+      <h3>Breach alert preferences</h3>
```
Comment (pdehaan) [inline src/views/partials/settings.js:44]: Q: Localized?

File: src/views/partials/settings.js [hunk 7]
```diff
@@ -1,5 +1,93 @@
...
+        <div id="add-email-modal-controls">
+          <label>${getMessage('email-address')}<input id="email" type="text"></label>
+          <button id="send-verification" class="monitor-button">${getMessage('send-verification')}</button>
+        </div>
+      </dialog>
+    </section>
+    <hr>
+    <section>
+      <h3>${getMessage('deactivate-account-title')}</h3>
+      <div>${getMessage('deactivate-account')}.</div>
```
Comment (pdehaan) [inline src/views/partials/settings.js:88]: Q: Should the period be within the translated string?

File: locales/en/settings.ftl [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+comm-opt-0 = Email me if one of my email addresses below appears in a data breach.
+comm-opt-1 = Send all breach alerts to { $primaryEmail }
+
+send-verification = Send Verification Link
+add-new-email = Add a new email address
+email-verification-required = Email verification required
+settings-title = { -product-short-name } Settings
+monitored-email-addresses = Monitored email addresses
+email-address-limit = Your account includes monitoring of up to { $limit } emails
+add-another-email = Add a new email address to see if it's been involved in a breach
```
[the remark this answers is not in the document]
Comment (flodolo) [reply locales/en/settings.ftl:24]: https://github.com/mozilla/blurts-server/pull/2763/checks

File: locales/en/settings.ftl [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+email-address = Email Address
+deactivate-account-title = Deactivate account
+deactivate-account = You can deactivate Firefox Monitor by deleting your Firefox account
```
[the remark this answers is not in the document]
Comment (flodolo) [reply locales/en/settings.ftl:27]: Indeed, both Firefox Monitor and Firefox account should use the corresponding term.

File: src/client/js/settings.js [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+      headers: {
+        'Content-Type': 'application/json',
+        'x-csrf-token': csrfToken
+      },
+      mode: 'same-origin',
+      method: 'POST',
+      body: JSON.stringify({ email })
+    })
+
+    if (!response || !response.ok) {
```
[the remark this answers is not in the document]
Comment (rhelmer) [reply src/client/js/settings.js:81]: This is from the existing front-end code, but I agree it doesn't make much sense as-is right now, will correct this.

**Comment by rhelmer** [comment]
This is still a draft PR, let's move this work over to https://github.com/mozilla/blurts-server/pull/2765 instead.

**Pull request closed without merging**